src/components/Viewport.ts

```diff
@@ -546,26 +660,66 @@ export default class Viewport {
     const panelManager = this.panelManager;
 
     // Restore index
-    panelManager.clear();
-    cameraElement.innerHTML = status.panels.map(panel => panel.html).join("");
-
-    this.createPanels();
-
-    // Reset panel index
-    panelManager.originalPanels().forEach((panel, idx) => {
-      panel.setIndex(panels[idx].index);
+    cameraElement.innerHTML = panels.map(panel => panel.html).join("");
+
+    // Create panels first
+    this.refreshPanels();
+    const createdPanels = panelManager.originalPanels();
+
+    // ...then order it by its index
+    const orderedPanels: Panel[] = [];
+    panels.forEach((panel, idx) => {
+      const createdPanel = createdPanels[idx];
+      createdPanel.setIndex(panel.index);
+      orderedPanels[panel.index] = createdPanel;
     });
-
-    this.currentPanel = panelManager.get(status.index)
-      || panelManager.get(defaultIndex)
-      || panelManager.firstPanel();
+    panelManager.replacePanels(orderedPanels, []);
+    panelManager.setCloneCount(0); // No clones at this point
+
+    const panelCount = panelManager.getPanelCount();
+    if (panelCount > 0) {
+      this.currentPanel = panelManager.get(status.index)
+        || panelManager.get(defaultIndex)
+        || panelManager.firstPanel();
+      this.nearestPanel = this.currentPanel;
+    } else {
+      this.currentPanel = undefined;
+      this.nearestPanel = undefined;
+    }
+    this.visiblePanels = orderedPanels.filter(panel => Boolean(panel));
 
     this.resize();
 
     this.axes.setTo({ flick: status.position }, 0);
     this.moveCamera(status.position);
   }
 
+  public calcVisiblePanels(): Panel[] {
+    const allPanels = this.panelManager.allPanels();
+    if (this.options.renderOnlyVisible) {
+      const visibleIndex = this.state.visibleIndex;
+      const getPanelFromRelativeIndex = (index: number) => {
+        index = index < 0
+          ? allPanels.length + index
+          : index;
+        return allPanels[index];
+      };
+
+      const visiblePanels = counter(visibleIndex.max - visibleIndex.min + 1).map(offset => {
+        const index = visibleIndex.min + offset;
+        return getPanelFromRelativeIndex(index);
+      }).filter(panel => panel);
```

Comment:
I hear the description of the calcVisiblePanels and think that I can modify the code as follows.


```js
const { min, max } = this.state.visibleIndex;
const backPanels = min >= 0 ? [] : allPanels.slice(min);
const frontPanels = allPanels.slice(Math.max(0, min), max + 1);

return [...frontPanels, ...backPanels].filter(panel => panel);
```